mesonbuild/interpreter.py

```diff
@@ -1883,10 +1883,14 @@ def func_project(self, node, args, kwargs):
             raise InvalidCode('Second call to project().')
         if not self.is_subproject() and 'subproject_dir' in kwargs:
             spdirname = kwargs['subproject_dir']
-            if '/' in spdirname or '\\' in spdirname:
-                raise InterpreterException('Subproject_dir must not contain a path segment.')
+            if not isinstance(spdirname, str):
+                raise InterpreterException('Subproject_dir must be a string')
+            if os.path.isabs(spdirname):
+                raise InterpreterException('Subproject_dir must not be an absolute path.')
             if spdirname.startswith('.'):
                 raise InterpreterException('Subproject_dir must not begin with a period.')
+            if '..' in spdirname:
```

Comment:
Note: please use `pathlib.PurePath().parts` so that names like `subprojects/some..dir/` don't fail this test.